Mathlib/Analysis/InnerProductSpace/JointEigenspace.lean

```diff
@@ -108,6 +111,131 @@ theorem directSum_isInteral_of_commute (hA : A.IsSymmetric) (hB : B.IsSymmetric)
+  · intro h
+    simp only [Submodule.mem_inf]
+    constructor
+    · simp only [Submodule.mem_map, Submodule.coeSubtype, Subtype.exists, exists_and_right,
+      exists_eq_right, mem_eigenspace_iff, SetLike.mk_smul_mk, restrict_apply,
+      Subtype.mk.injEq] at h
+      obtain ⟨_, hy⟩ := h
+      simpa [mem_eigenspace_iff]
+    · simp only [Submodule.coeSubtype] at h
+      obtain ⟨_, hy⟩ := h
+      simp only [← hy.2, Submodule.coeSubtype, SetLike.coe_mem]
+
+open Classical
+
+/--The orthocomplement of the indexed supremum of joint eigenspaces of a finite commuting tuple of
+symmetric operators is trivial. -/
+theorem orthogonalComplement_iSup_iInf_eigenspaces_eq_bot [Fintype n] [FiniteDimensional 𝕜 E]
+    (T : n → (E →ₗ[𝕜] E)) (hT :(∀ (i : n), ((T i).IsSymmetric)))
+    (hC : (∀ (i j : n), (T i) ∘ₗ (T j) = (T j) ∘ₗ (T i))) :
+    (⨆ (γ : n → 𝕜), (⨅ (j : n), (eigenspace (T j) (γ j)) : Submodule 𝕜 E))ᗮ = ⊥ := by
+  revert T
+  refine Fintype.induction_subsingleton_or_nontrivial n ?_ ?_
+  · intro m _ hhm T hT _
+    simp only [Submodule.orthogonal_eq_bot_iff]
+    by_cases case : Nonempty m
+    · have i := choice case
+      have := uniqueOfSubsingleton i
+      conv => lhs; rhs; ext γ; rw [ciInf_subsingleton i]
+      rw [← (Equiv.funUnique m 𝕜).symm.iSup_comp]
+      apply Submodule.orthogonal_eq_bot_iff.mp ((hT i).orthogonalComplement_iSup_eigenspaces_eq_bot)
+    · simp only [not_nonempty_iff] at case
+      simp only [iInf_of_empty, ciSup_unique]
+  · intro m hm hmm H T hT hC
+    obtain ⟨w, i , h⟩ := exists_pair_ne m
+    simp only [ne_eq] at h
+    have D := H {x // x ≠ i} (Fintype.card_subtype_lt (p := fun (x : m) ↦ ¬x = i) (x := i)
+      (by simp only [not_true_eq_false, not_false_eq_true])) (Subtype.restrict (fun x ↦ x ≠ i) T)
+        (fun (i_1 : {x // x ≠ i}) ↦ hT ↑i_1) (fun (i_1 j : { x // x ≠ i }) ↦ hC ↑i_1 ↑j)
+    simp only [Submodule.orthogonal_eq_bot_iff] at *
+    have G : (⨆ (γ : {x // x ≠ i} → 𝕜), (⨆ μ : 𝕜, (eigenspace (T i) μ ⊓ (⨅ (j : {x // x ≠ i}),
+    eigenspace (Subtype.restrict (fun x ↦ x ≠ i) T j) (γ j))))) = ⨆ (γ : {x // x ≠ i} → 𝕜),
+    (⨅ (j : {x // x ≠ i}), eigenspace (Subtype.restrict (fun x ↦ x ≠ i) T j) (γ j)) := by
+      conv => lhs; rhs; ext γ; rhs; ext μ; rw [invariant_subspace_inf_eigenspace_eq_restrict (T i) μ
+        (iInf_eigenspace_invariant_of_commute T hC i γ)]
+      conv => lhs; rhs; ext γ; rw [iSup_simultaneous_eigenspaces_eq_top (T i) (hT i)
+        (iInf_eigenspace_invariant_of_commute T hC i γ)]
+    have H1 : ∀ (i : m), ∀ (s : m → 𝕜 → Submodule 𝕜 E), (⨆ f : m → 𝕜, ⨅ x, s x (f x)) =
+        ⨆ f' : {y // y ≠ i} → 𝕜, ⨆ y : 𝕜, s i y ⊓ ⨅ x' : {y // y ≠ i}, (s x' (f' x')) := by
+      intro i s
+      rw [← (Equiv.funSplitAt i 𝕜).symm.iSup_comp, iSup_prod, iSup_comm]
+      congr! with f' y
+      rw [iInf_split_single _ i, iInf_subtype]
+      congr! with x hx
+      · simp
+      · simp [dif_neg hx]
+    rw [← G] at D
+    rw [H1 i (fun _ ↦ (fun μ ↦ (eigenspace (T _) μ )))]
+    exact D
```

Comment:
This is how I would clean this up a bit:

1. use `isEmpty_or_nonempty` instead of `by_cases`; often the bespoke `∨` statements are nicer than using `by_cases` because you don't have to `simp` the negation through the statement.
2. No need to give a name to `i`.
3. using `exists_pair_ne` is overkill because you don't need `w` at all, so using `Classical.arbitrary` to get `i` works.
4. Instead of `have D`, you can just `specialize H`, and that prevents cluttering the infoview.
5. Instead of `have G` and later `rw [← G] at H`, you can just do `conv at H` to rewrite  `H` on the spot.
6. By pulling the `rw [← (Equiv.funSplitAt i 𝕜).symm.iSup_comp, iSup_prod, iSup_comm]` out into the main proof, we get that `H` and `⊢` match very closely, so applying `convert` is nice, and it can simultaneously introduce variables for us.
7. As I predicted above, we don't actually need the earlier lemma `invariant_submodule_inf_eigenspace`.

Note: doing it this way exposed an API hole: we don't have a `genEigenspace_one` lemma.
```suggestion
  refine Fintype.induction_subsingleton_or_nontrivial n (fun m _ hhm T hT _ ↦ ?_)
    (fun m hm hmm H T hT hC ↦ ?_)
  · obtain (hm | hm) := isEmpty_or_nonempty m
    · simp
    · have := uniqueOfSubsingleton (choice hm)
      simpa only [ciInf_unique, ← (Equiv.funUnique m 𝕜).symm.iSup_comp]
        using hT default |>.orthogonalComplement_iSup_eigenspaces_eq_bot
  · have i := Classical.arbitrary m
    specialize H {x // x ≠ i} (Fintype.card_subtype_lt (x := i) (by simp))
      (Subtype.restrict (· ≠ i) T) (hT ·) (hC · ·)
    simp only [Submodule.orthogonal_eq_bot_iff] at *
    conv at H =>
      enter [1, 1, γ]
      rw [← iSup_eigenspace_restrict (T i) (hT i) (iInf_eigenspace_invariant_of_commute hC i γ)]
      enter [1, μ]
      -- we should add a `genEigenspace_one` lemma, maybe marked `@[simp]`.
      change map (Submodule.subtype _)
        (genEigenspace ((T i).restrict <| iInf_eigenspace_invariant_of_commute hC i γ) μ 1)
      rw [← Submodule.inf_genEigenspace _ _ _ (k := 1), inf_comm]
    rw [← (Equiv.funSplitAt i 𝕜).symm.iSup_comp, iSup_prod, iSup_comm]
    convert H with γ μ
    rw [iInf_split_single _ i, iInf_subtype]
    congr! with x hx
    · simp; rfl -- `rfl` isn't necessary with a `genEigenspace_one` lemma
    · simp [dif_neg hx]
```

Reply:
There's also this, starting from my line 185 `conv at H`, replacing that and everything below with:
```lean
    rw [← (Equiv.funSplitAt i 𝕜).symm.iSup_comp, iSup_prod, iSup_comm]
    convert H with γ
    rw [← iSup_eigenspace_restrict (T i) (hT i) (iInf_eigenspace_invariant_of_commute hC i γ)]
    congr! with μ
    -- need a `genEigenspace_one` lemma
    change _ = map (Submodule.subtype _)
      (genEigenspace ((T i).restrict <| iInf_eigenspace_invariant_of_commute hC i γ) μ 1)
    rw [← Submodule.inf_genEigenspace _ _ _ (k := 1), inf_comm, iInf_split_single _ i, iInf_subtype]
    congr! with x hx
    · simp; rfl -- `rfl` isn't necessary with a `genEigenspace_one` lemma
    · simp [dif_neg hx]
```

Reply:
This one is probably substantial...super helpful though. I'll have to bang my head against this a bit.